components/script/dom/htmlimageelement.rs

```diff
@@ -110,6 +110,60 @@ impl Runnable for ImageResponseHandlerRunnable {
     }
 }
 
+/*/// The context required for asynchronously loading an external image.
+struct ImageContext {
+    /// The element that initiated the request.
+    elem: Trusted<HTMLImageElement>,
+    /// The response body received to date.
+    data: Vec<u8>,
+    /// The response metadata received to date.
+    metadata: Option<Metadata>,
+    /// The initial URL requested.
+    url: Url,
+    /// Indicates whether the request failed, and why
+    status: Result<(), NetworkError>
+}
+
+impl AsyncResponseListener for ImageContext {
+    fn headers_available(&mut self, metadata: Result<Metadata, NetworkError>) {
+        self.metadata = metadata.ok();
+
+        let status_code = self.metadata.as_ref().and_then(|m| {
+            match m.status {
+                Some(RawStatus(c, _)) => Some(c),
+                _ => None,
+            }
+        }).unwrap_or(0);
+
+        self.status = match status_code {
+            0 => Err(NetworkError::Internal("No http status code received".to_owned())),
+            200...299 => Ok(()), // HTTP ok status codes
+            _ => Err(NetworkError::Internal(format!("HTTP error code {}", status_code)))
+        };
+    }
+
+    fn data_available(&mut self, payload: Vec<u8>) {
+        if self.status.is_ok() {
+            let mut payload = payload;
+            self.data.append(&mut payload);
+        }
+    }
+
+    fn response_complete(&mut self, status: Result<(), NetworkError>) {
+        let load = status.and(self.status.clone()).map(|_| {
+            let data = mem::replace(&mut self.data, vec!());
+            let metadata = self.metadata.take().unwrap();
```

Comment:
It's at this point that we will want to use the new `StoreDecodeImage` command to the image cache with the bytes from `self.data`.